Add metalk8s-sosreport to the debian packages

ycointe opened the pull request:
**Component**: packages

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: Metalk8s over Ubuntu

**Summary**: Need to add sources to build metalk8s-sosreport

**Acceptance criteria**: 
Be able to build metalk8s-sosreport:

- Move into metalk8s-sosreport package forlder: ` $ cd packages/debian/metalk8s-sosreport-deb-2.4.0`
- Build the package: ` $ debuild -b -uc -us`
 -> You should have `metalk8s-sosreport-deb_2.4.0_amd64.deb` in the parent folder.

You can verify if the package is working well:
- Download and install sosreport: ` $ sudo apt install sosreport`
- Install metalk8s-sosreport: ` $ dpkg -i metalk8s-sosreport-deb_2.4.0_amd64.deb `
-> then verify the installation: ` $ sosreport -l | grep -E "metalk8s|containerd" `

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1471 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

ycointe committed:
Add metalk8s-sosreport to the debian packages

Add the sources to build to obtain metalk8s-sosreport
to the debian packages repository

bert-e commented:
# Hello ycointe,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @gdemonet

* @monpote

* @sayf-eddine-scality

* @alexandre-allard-scality

* @zempashi

* @carlito-scality

* @lucieleonard

* @thomasdanan

* @teddyandrieux

* @nootal

* @cmonfort

* @slaperche-scality

* @nicolast

* @jbertran

sayf-eddine-scality commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/changelog:
```diff
@@ -0,0 +1,12 @@
+
+ -- Yohann Cointe <yohann.cointe@scality.com>  Wed, 24 Jul 2019 17:11:02 +0200
+
+metalk8s-sosreport (2.4.0) UNRELEASED; urgency=low
```
useless input, only keep latest changelog

sayf-eddine-scality commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/control:
```diff
@@ -0,0 +1,13 @@
+Source: metalk8s-sosreport
+Section: base
+Priority: extra
+Maintainer: Yohann Cointe <yohann.cointe@scality.com>
```
maintainer should be Moonshot team, for future support
(changelog) too

sayf-eddine-scality commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/control:
```diff
@@ -0,0 +1,13 @@
+Standards-Version: 2.4.0 
+Homepage: https://github.com/scality/metalk8s
+
+Package: metalk8s-sosreport-deb
```
no need to `-deb` as the file literally ends with `.deb`

sayf-eddine-scality commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/metalk8s-sosreport-deb/DEBIAN/control:
```diff
@@ -0,0 +1,12 @@
+Package: metalk8s-sosreport-deb
```
dublicate file ? one of them should be unnecessary

sayf-eddine-scality commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/metalk8s-sosreport-deb/usr/share/sosreport/sos/plugins/containerd.py:
```diff
@@ -0,0 +1,73 @@
+#! /bin/env python
```
I guess you did a `git add -A` some files are generated on build time and do not need to be in the cvs

sayf-eddine-scality commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/metalk8s-sosreport-deb/usr/share/sosreport/sos/plugins/metalk8s.py:
```diff
@@ -0,0 +1,119 @@
+#! /bin/env python
```
same

sayf-eddine-scality commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/src/containerd.py:
```diff
@@ -0,0 +1,73 @@
+#! /bin/env python
```
so this file and the next are common to debian and redhat, they should reside in one place to ease maintenance and should be copied (linked) in build time.

ycointe replied:
Is there no chance that these files could evolve with the time ?
If one file change for an OS family and not for another, it could be a problem if different OS families use the same.

sayf-eddine-scality replied:
good point but I am not sure we want to support two codebases or have one file that supports both OSs, tbd

slaperche-scality replied:
Please, don't use `python`: use either `python2` or `python3`
Only use `python` when your code support both version.

I've had to patch too many scripts that were broken on ArchLinux because they now use Python 3 as default but people still think `python` == `python2` which is wrong (see [PEP-394](https://www.python.org/dev/peps/pep-0394/))

ycointe replied:
As python3 will be the next standard and will be installed by default on Debian systems, I suggest that we only use python3.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @gdemonet

* @monpote

* @sayf-eddine-scality

* @alexandre-allard-scality

* @zempashi

* @carlito-scality

* @lucieleonard

* @thomasdanan

* @teddyandrieux

* @nootal

* @cmonfort

* @slaperche-scality

* @nicolast

* @jbertran





The following reviewers are expecting changes from the author, or must review again:

* @sayf-eddine-scality

slaperche-scality commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/copyright:
```diff
@@ -0,0 +1,9 @@
+Upstream-Name: metalk8s-sosreport
+Source: <url://example.com>
+#
+# Please double check copyright with the licensecheck(1) command.
```
I guess we will want to use the Apache License (which is what we use for the project).

slaperche-scality commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/src/containerd.py:
```diff
@@ -0,0 +1,73 @@
+from sos.plugins import Plugin, RedHatPlugin, UbuntuPlugin
+
+
+class containerd(Plugin, RedHatPlugin, UbuntuPlugin):
```
If we do keep two versions of the code, then I think the Ubuntu version doesn't need to inherit from `RedHatPlugin` and the CentOS version doesn't need to inherit from `UbuntuPlugin`.

But seeing the quantity of duplicated code, I think we will want to keep a single implementation that support both OSes.

Worst case (if they do diverge a lot) we could have something like:

```python
class UbuntuContainerd(Plugin, UbuntuPlugin):
    …

class RedHatContainerd(Plugin, RedHatPlugin):
    …

if os_is_ubuntu:
     containerd = UbuntuContainerd
elif os_is_redhat:
     containerd = RedHatContainerd
else:
    raise some_error
```

but that won't be needed for now.

jbertran replied:
In the meantime, this is a possibility for inheriting from the correct plugin:

```
In [1]: class Both: 
   ...:     pass 
   ...:                                                             

In [2]: class One: 
   ...:     def a(self): 
   ...:         print('toto') 
   ...:                                                             

In [3]: class Other: 
   ...:     def a(self): 
   ...:         print('tata') 
   ...:                                                             

In [4]: selector = True                                             

In [5]: class Inheritor(Both, One if selector else Other): 
   ...:     pass 
   ...:                                                             

In [6]: i = Inheritor()                                             

In [7]: i.a()                                                       
toto
```

slaperche-scality commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/metalk8s-sosreport-deb.substvars:
```diff
@@ -0,0 +1,2 @@
+misc:Depends=
+misc:Pre-Depends=
```
Don't wee need to depends on some version of python and `sos`?

jbertran replied:
This is `substvars`, we don't actually need this file to have contents (or at all???) if we fill up what we need in `control`.

@ycointe please check whether or not we need this file to be present at all

jbertran commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/.debhelper/generated/metalk8s-sosreport-deb/installed-by-dh_install:
```diff
@@ -0,0 +1,2 @@
+./debian/src/metalk8s.py
```
I don't think these `.debhelper` generated files should be committed, should they?

jbertran commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/README.Debian:
```diff
@@ -0,0 +1,8 @@
+metalk8s-sosreport for Debian
+
+Please edit this to provide information specific to 
```
Let's have a significant description

jbertran commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/README.Debian:
```diff
@@ -0,0 +1,8 @@
+
+    (Automatically generated by debmake Version 4.2.9)
+
+ -- Yohann <ycointe@localhost>  Wed, 24 Jul 2019 15:49:02 +0200
```
Let's generate these files with the correct email address for you.

See [here](https://people.debian.org/~osamu/maint-guide.html#email-setup) for that.

ycointe replied:
Thx !

jbertran commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/control:
```diff
@@ -0,0 +1,13 @@
+Source: metalk8s-sosreport
+Section: base
+Priority: extra
+Maintainer: Moonshot Team <moonshot-platform@scality.com>
```
Platform team

jbertran commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/copyright:
```diff
@@ -0,0 +1,9 @@
+Format: http://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
+Upstream-Name: metalk8s-sosreport
+Source: <url://example.com>
```
Stray example URL

jbertran commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/watch:
```diff
@@ -0,0 +1,2 @@
+# You must remove unused comment lines for the released package.
```
Let's follow the instructions :)

jbertran reviewed (commented):
I'm by no means an expert on debian packaging, ignore at will if you have better information.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @sayf-eddine-scality

jbertran commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/README.Debian:
```diff
@@ -0,0 +1,5 @@
+metalk8s-sosreport for Debian
+
+Metalk8s sos-report custom plugins
```
Let's be consistent between OS for the naming.

It's `Metalk8s SOS report custom plugins` in the `.spec`.

jbertran commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/rules:
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/make -f
+# You must remove unused comment lines for the released package.
```
Let's follow the instructions :)

jbertran commented on packages/debian/metalk8s-sosreport-deb-2.4.0/debian/source/local-options:
```diff
@@ -0,0 +1,2 @@
+#abort-on-upstream-changes
```
I'm guessing we can remove this file if we keep these options commented. From [here](https://www.debian.org/doc/manuals/maint-guide/dother.en.html#sourcel) it looks like these are for a specific workflow which is not our own - even if we will need that file later, we can add it back when we do the buildchain integration.

jbertran reviewed (commented):
We can still trim it down a little :)

Let's also think about not having the custom plugin code twice in git.

ycointe committed:
Add a directory for common files

Move common files for metalk8s-sosreport into the common directory.

slaperche-scality commented on packages/common/containerd.py:
```diff
@@ -0,0 +1,73 @@
+#! /bin/env python
```
Please specify either `python2` or `python3` ditto for the other file.

Also, about the location, I would put the files under `packages/common/metalk8s-sosreport` (or even `packages/common/metalk8s-sosreport/src`).

In case we have more packages in the future we shouldn't put stuff directly under `common`.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @sayf-eddine-scality

* @slaperche-scality

ycointe committed:
Change RedHat's metalk8s-sosreport sources path

Change the metalk8s-sosreport sources path to use
the sources in the common folder.
So need to delete the previous sources in the redhat folder
to avoid having the same files multiple times in multiple locations.

jbertran commented on packages/common/metalk8s-sosreport/containerd.py:
```diff
@@ -1,4 +1,4 @@
-#! /bin/env python
```
Can we have a separate commit for these changes ?

jbertran commented on packages/debian/README.md:
```diff
@@ -0,0 +1,31 @@
+## Requirement
+
+To build mannually Debian packages, install `devscripts`, `debhelper`
+and `buid-essential`:
```
Typo, and let's have the commands with a helpful package name since we don't all run Ubuntu.

`to build manually, the commands dh and debuild are required on top of the MetalK8s build requirements.` or something like that

jbertran commented on packages/debian/README.md:
```diff
@@ -0,0 +1,31 @@
+
+## Requirement
+
+To build mannually Debian packages, install `devscripts`, `debhelper`
```
`s/manually/manually`

jbertran reviewed (commented):
Lookin' good! Minor stuff to amend :)

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/1471-sosreport-package-creation` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `improvement/1471-sosreport-package-creation`
* Rebase `improvement/1471-sosreport-package-creation` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

slaperche-scality commented on packages/debian/README.md:
```diff
@@ -0,0 +1,31 @@
+<h1>
```
Is it needed?

ycointe replied:
It's not mandatory but it can be useful for now.

gdemonet replied:
Let's not start littering the repo with `.md` files again. We have real documentation, let's use it please :)

jbertran replied:
This is dev only and only for a very specific purpose though (building packages by hand).

gdemonet replied:
> This is dev only 

`docs/developer`

> and only for a very specific purpose though (building packages by hand).

We have other examples of temporary docs in there as well (accessing services without an Ingress for instance).

Also, building packages by hand may be a valid use-case that we'll want to keep in the future for dev purposes.

ycointe replied:
The file will be kept until it is automated.
When it will be automated, the .md content file will be integrated in the documentation and will be deleted.
Does everyone is agree ?

slaperche-scality replied:
Agree. What we want to document is how to build package using a container +  `entrypoint.sh`

This is only temporary note until we're here.

gdemonet commented on packages/debian/metalk8s-sosreport/debian/control:
```diff
@@ -0,0 +1,13 @@
+Source: metalk8s-sosreport
+Section: base
+Priority: extra
+Maintainer: Plateform Team <moonshot-platform@scality.com>
```
```suggestion
Maintainer: Scality Platform Team <moonshot-platform@scality.com>
```

gdemonet commented on packages/debian/metalk8s-sosreport/debian/copyright:
```diff
@@ -0,0 +1,12 @@
+Source: <https://github.com/scality/metalk8s>
+#----------------------------------------------------------------------------
+#
+# 		    GNU GENERAL PUBLIC LICENSE
```
No I don't think we want to use `GPL` licenses. Let's stick with the same licensing policy as for the parent repo, Apache v2 (cf. [here](https://github.com/scality/metalk8s/blob/development/2.4/LICENSE)).

slaperche-scality replied:
Depends if SOS is GPL and requires its plugin to be GPL as well.

To be checked, but yeah if possible let's stick to a single license.

ycointe replied:
I checked, SOS is under GPL v2 licence.
Do we keep this licence or do we keep the Apache licence ?

NicolasT replied:
This can be Apache2'd just like other parts of MetalK8s.

ycointe replied:
I used the Apache 2 licence.
Thank you for your help about it.

gdemonet commented on packages/debian/metalk8s-sosreport/debian/src/containerd.py:
```diff
@@ -0,0 +1,73 @@
+#! /bin/env python3
```
Don't think we still need these, since they also are under a `common` directory, right?

slaperche-scality replied:
Why should we remove the shebang?

ycointe replied:
These files (containerd.py and metalk8s.py) should have been deleted after tests for the common folder.
I will take care of it.

gdemonet replied:
(not related to the shebang sorry)

jbertran committed:
sosreport: make the shebang portable

Fix the python version to `python3` instead of relying on aliases for
`python`.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @sayf-eddine-scality

* @slaperche-scality

* @gdemonet

ycointe commented:
@bert-e approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @sayf-eddine-scality

* @slaperche-scality

* @gdemonet







*The following options are set:* **approve**

jbertran commented:
Everything square after deployment, all comments taken into account.

Nice work! Now for the real fun - calico, and the buildchain next :smile:

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @sayf-eddine-scality

* @slaperche-scality







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @sayf-eddine-scality







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye ycointe.
